lib/api/registration/RegistrationSourceLocation.ts

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright © 2019 Atomist, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as path from "path";
+import * as trace from "stack-trace";
+/**
+ * Where is this registration located in the code?
+ * Populate this with guessSourceLocation()
+ *
+ * This lets an autofix commit message show where the autofix was registered,
+ * giving developers the information they need to change it.
+ */
+export interface RegistrationSourceLocation {
+    url: string;
+}
+
+/**
+ * Use within declaration of an AutofixRegistration:
+ * {
+ * ...
+ * registrationSourceLocation: guessSourceLocation()
+ * }
+ */
+export function guessSourceLocation(): RegistrationSourceLocation | undefined {
+    try {
+
+        const fileLocation = guessLocationWithinProject();
+        const projectLocation = guessProjectLocation();
+
+        return {
+            url: guessUrl(projectLocation, fileLocation),
+        };
+    } catch (err) {
+        return undefined;
+    }
+}
+
+interface GuessedLocationWithinProject {
+    relativePath: string; lineNumber: number;
+}
+function guessLocationWithinProject(): GuessedLocationWithinProject {
+    let stack: trace.StackFrame[];
+    try {
+        // Just throw an error so that we can capture the stack
+        throw new Error();
+    } catch (err) {
+        stack = trace.parse(err);
+    }
+    // 0 = guessLocationWithinProject, 1 = guessSourceLocation, 2 = the caller, where it's registered
+    const registration = stack[2];
+
+    const appRoot = process.cwd();
+    const relativePath = registration.getFileName().replace(appRoot, "");
+    return {
+        relativePath,
+        lineNumber: registration.getLineNumber(),
+    };
+}
+
+interface GuessedProjectLocation { repoUrl: string; sha?: string; }
+function guessProjectLocation(): GuessedProjectLocation {
+
+    try {
+        const gitInfo: { sha: string, repository: string } =
+            require(process.cwd() + path.sep + "git-info.json");
```

Comment:
process.cwd won’t likely work inside docker. Please use appRoot.path. Additionally use path.join to construct the path.

Reply:
path.join, agree; appRoot.path didn't work, it found the root of the library not the project

Reply:
there are a lot of cases here that I didn't want to put a ton of work into before asking your opinion via this PR.

Like, when the registration is created in a library, show the lib code or the top level?
... hmm moving this into the addRegistration methods might help. although not with the apply-pack methods. Still, those are less common. Maybe we need to put the name of the SDM there too.